src/anyp/Uri.cc

```diff
@@ -937,55 +947,62 @@ AnyP::Uri::Uri(AnyP::UriScheme const &aScheme) :
 }
 
 // TODO: fix code duplication with AnyP::Uri::parse()
-char *
-AnyP::Uri::cleanup(const char *uri)
+SBuf
+AnyP::Uri::Cleanup(SBuf &uri)
 {
+    // 'whitespace' for chop and strip are different
+    static const CharacterSet wspChop = (CharacterSet::WSP +
+                                         CharacterSet::LF +
+                                         CharacterSet::CR).rename("wsp-chop");
+    static const CharacterSet wspStrip = (CharacterSet("VT,FF","\v\f") +
+                                          wspChop).rename("wsp-strip");
+
     int flags = 0;
-    char *cleanedUri = nullptr;
+    SBuf cleanedUri;
     switch (Config.uri_whitespace) {
     case URI_WHITESPACE_ALLOW:
         flags |= RFC1738_ESCAPE_NOSPACE;
     // fall through to next case
     case URI_WHITESPACE_ENCODE:
         flags |= RFC1738_ESCAPE_UNESCAPED;
-        cleanedUri = xstrndup(rfc1738_do_escape(uri, flags), MAX_URL);
+        cleanedUri = SBuf(rfc1738_do_escape(uri.c_str(), flags));
```

Comment:
Intentional, the MAX_URL truncation is/was done in the existing callers except for error page display. Future callers do not have to obey that restriction if they support SBuf URI's.

Reply:
> it is safe. The MAX_URL truncation is/was done in the existing callers except for error page display.

I see no immediate MAX_URL truncation in the two existing callers of this Uri::Cleanup() function:

```
absorbLogUri(SBufToCstring(AnyP::Uri::Cleanup(canonicalUri)));
al->setVirginUrlForMissingRequest(AnyP::Uri::Cleanup(rawUriBuf));
```

It is, of course, possible that all URLs being fed to Uri::Cleanup() were and will continue to be properly truncated before the function is called and/or that all URIs returned by Cleanup() were and will continue to be properly truncated later, as needed. This is essentially what you assert.

FWIW, I am not convinced by your assertion because I see no reasonable way to validate it and to make sure that it remains valid across future code changes. I believe the correct way to get rid of MAX_URL is to remove it in one comprehensive PR. Certainly, there are easy-to-track cases where specific MAX_URL occurrences can and should be removed in isolation, but, IMO, this is not one of those cases.

I am resolving this change request because you seem to be sure that this MAX_URL removal is safe, and I have no reasonable way to invalidate your assertion, but I still recommend adding code to preserve the old MAX_URL limit here (until it is removed in a comprehensive PR mentioned above).